src/kritter/ktextvisor.py

```diff
@@ -0,0 +1,298 @@
+            if words[0].lower()!="help":
+                claimed = bool([r.claim for r in responses if r.claim])
+                if not claimed:
+                    responses += [Response('Try "help".')]
+            self.send_responses(sender, responses)
+
+    def close(self):
+        try:
+            self.server.stop_event.set()
+            self.thread.join()
+        except:
+            pass
+
+    def callback_receive(self, prepend=False):
+        def wrap_func(func):
+            if prepend:
+                self.callbacks.insert(0, func)
+            else:
+                self.callbacks.append(func)
+        return wrap_func
+
+    def call_callbacks(self, words, sender, context):
+        responses = []
+        for c in self.callbacks.copy():
+            try:
+                response = c(words, sender, context)
+                if response==[]:
+                    continue
+                elif isinstance(response, Response):
+                    responses += [response]
+                elif isinstance(response, list) and isinstance(response[0], Response):
+                    responses += response
+                elif response:
+                    responses += [Response(response)]
+                claimed = bool([r.claim for r in responses if r.claim])
+                if claimed:
+                    break
+            except ConnectionRefusedError:
+                self.callbacks.remove(c)
+        return responses
+    
+    def pre_handle_context(self, sender):
+        t = time.time()
+        # c is a tuple: (timestamp, tags):
+        self.context = {_sender: c for _sender, c in self.context.items() if t-c[0]<CONTEXT_TIMEOUT}
+        try:
+            return self.context[sender['id']][1]
+        except:
+            return []
+
+    def post_handle_context(self, sender, responses):
+        t = time.time()
+        contexts = [r.context for r in responses]
+        contexts = [c for c in contexts if c is not None]
+        if len(contexts):
+            context = set(contexts[-1])
+        else:
+            try:
+                context = self.context[sender['id']][1]
+            except:
+                context = []
+        if context==[]:
+            try:
+                del self.context[sender['id']]
+            except: 
+                pass
+        else:
+            self.context[sender['id']] = (t, context)
+
+                        
+    def send_responses(self, sender, responses):
+        content = [r.content for r in responses] # extract content
+        content = format_content(content) # combine strs and dicts
+        content = [format_dict(c) if isinstance(c, dict) else c for c in content] # format dicts into strs
+        content = format_content(content) # combine strs again
+        with self.lock:
+            self.text_client.send(content, sender)
+
+    def native_callback(self, words, sender, context):
+        # userids need to be saved as strings to avoid
```

Comment:
userid's are unique, yes?  Why do you need to make them strings?  
I don't think you need to call config.load().  You load it in __init__ (why would it change?)

Reply:
config.load() -- removed. I thought there would be an data mismatch from having the users removed in the texting dialog

unique userIds -- an issue of duplicates from different types. When json.loads() returns, the dict in config['subscribers'] will have casted the integer to a string. Users will be allowed to subscribe twice, once with their id as an integer and once with their id as a string. Ex -- { 10 : 'me', '10' : 'me' }.. an alternative would be casting the config dictionary keys upon load.